### PR TITLE
ESQL: Fix IpPrefix function not handling correctly ByteRefs

### LINKED ISSUE #109198: [CI] CsvTests test {ip.IpPrefixLengthFromColumn} failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/cusmwrqqsgnfu/tests/:x-pack:plugin:esql:test/org.elasticsearch.xpack.esql.CsvTests/test%20%7Bip.IpPrefixLengthFromColumn%7D

**Reproduction line:**
```
./gradlew ':x-pack:plugin:esql:test' --tests "org.elasticsearch.xpack.esql.CsvTests" -Dtests.method="test {ip.IpPrefixLengthFromColumn}" -Dtests.seed=216F39F857BEB359 -Dtests.jvm.argline="-Des.concurrent_search=true" -Dtests.locale=pt-PT -Dtests.timezone=Asia/Chongqing -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
[Failure dashboard for `org.elasticsearch.xpack.esql.CsvTests#test {ip.IpPrefixLengthFromColumn}`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('test%20%7Bip.IpPrefixLengthFromColumn%7D'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.CsvTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
java.lang.AssertionError: Data mismatch:
row 1 column 2:a list containing
row 1 column 2:0: expected "::" but was "::1"

  at __randomizedtesting.SeedInfo.seed([216F39F857BEB359:A93B0622F942DEA1]:0)
  at org.elasticsearch.xpack.esql.CsvTests.ip.IpPrefixLengthFromColumn(ip.csv-spec:540)
  at org.junit.Assert.fail(Assert.java:89)
  at org.elasticsearch.xpack.esql.CsvAssert.dataFailure(CsvAssert.java:260)
  at org.elasticsearch.xpack.esql.CsvAssert.assertData(CsvAssert.java:250)
  at org.elasticsearch.xpack.esql.CsvAssert.assertData(CsvAssert.java:170)
  at org.elasticsearch.xpack.esql.CsvAssert.assertData(CsvAssert.java:160)
  at org.elasticsearch.xpack.esql.CsvAssert.assertResults(CsvAssert.java:52)
  at org.elasticsearch.xpack.esql.CsvTests.assertResults(CsvTests.java:273)
  at org.elasticsearch.xpack.esql.CsvTests.doTest(CsvTests.java:264)
  at org.elasticsearch.xpack.esql.CsvTests.test(CsvTests.java:242)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1570)

```

### PR DESCRIPTION
**Pull request by ivancea**
- Fix IpPrefix function not using ByteRef offset on calculations, leading to wrong bytes being written to the prefix.
- Fixed the CSVTest case, which was also wrong, leading to a "_wrong * wrong = correct_" situation.
-  ~~Added` a BytesRef randomizer and applied it to the IpPrefixTests suppliers~~
    - ~~This highlighted another bug, which was fixed too~~
    - (Moved to https://github.com/elastic/elasticsearch/pull/109207)

Fixes https://github.com/elastic/elasticsearch/issues/109198

### CODE REVIEW

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefixTests.java [hunk 1]
```diff
@@ -106,7 +106,7 @@ public static Iterable<Object[]> parameters() {
             })
         );
 
-        return parameterSuppliersFromTypedData(errorsForCasesWithoutExamples(anyNullIsNull(true, suppliers)));
+        return parameterSuppliersFromTypedData(errorsForCasesWithoutExamples(anyNullIsNull(true, randomizeBytesRefsOffset(suppliers))));
```
Comment (ivancea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefixTests.java:109]: This chain is starting to (And will) get longer over time I think. Should we replace it with a "recommendedModifiers()"? Specially for average cases where all of them are applied.
Maybe not for this PR though, as this is a fix

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java [hunk 2]
```diff
@@ -642,6 +642,46 @@ public interface ExpectedEvaluatorToString {
         Matcher<String> evaluatorToString(int nullPosition, TestCaseSupplier.TypedData nullData, Matcher<String> original);
     }
 
+    /**
+     * Modifies suppliers to generate BytesRefs with random offsets.
+     */
+    protected static List<TestCaseSupplier> randomizeBytesRefsOffset(List<TestCaseSupplier> testCaseSuppliers) {
```
Comment (ivancea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:648]: Wasn't sure about pushing this in this PR, but as it discovered another bug, well...

My idea would be to propagate it to other tests (Or at least to new functions)

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefix.java [hunk 3]
```diff
@@ -136,7 +136,14 @@ static BytesRef process(
             throw new IllegalArgumentException("Prefix length v6 must be in range [0, 128], found " + prefixLengthV6);
         }
 
-        boolean isIpv4 = Arrays.compareUnsigned(ip.bytes, 0, IPV4_PREFIX.length, IPV4_PREFIX, 0, IPV4_PREFIX.length) == 0;
+        boolean isIpv4 = Arrays.compareUnsigned(
+            ip.bytes,
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefix.java:140]: Oh good catch. Lots of  `BytesRef`s do indeed start at 0, but the ones we use might not.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:648]: :+1: on separating it it for now.

I wonder if we could merge this into `TypedData`'s constructor. It already has a funny hack for `UNSIGNED_LONG`. I wonder if we can do a hack for `BytesRef` and randomize the offset there.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java [hunk 4]
```diff
@@ -642,6 +642,45 @@ public interface ExpectedEvaluatorToString {
...
+
+            var newData = testCase.getData().stream().map(typedData -> {
+                if (typedData.data() instanceof BytesRef bytesRef) {
+                    var offset = randomIntBetween(0, 10);
+                    var newBytesArray = new byte[bytesRef.length + offset];
+
+                    System.arraycopy(bytesRef.bytes, bytesRef.offset, newBytesArray, offset, bytesRef.length);
+
+                    var newBytesRef = new BytesRef(newBytesArray, offset, bytesRef.length);
+                    var newTypedData = new TestCaseSupplier.TypedData(newBytesRef, typedData.type(), typedData.name());
```
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:660]: While you are here you could randomly add trailing bytes too. You could fill them with random garbage too if you were especially adventurous.

**Review by nik9000 — APPROVED** [review]
I think we should merge this one as is and then iterate on the trick to randomize the offsets in the bytesrefs in a follow up chagne.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java (hunk 2 above)
Comment (ivancea) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:648]: Ok! Extracting to another branch, and merging

Comment (ivancea) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java]: About the TypedData constructor, I would avoid it, as I could imagine some test manually adding offsets themselves for some usecase (?)
Not sure, but I feel like there's more control this way

**Pull request merged**